DAOS-13697 doc: add document for pool upgrade (#12520)
daos-stack/daos: docs/admin/pool_operations.md

Additional status and telemetry data is planned to be exported through
management tools and will be documented here once available.

### Upgrading a Pool

> The pool upgrade operation upgrades pool's disk format to latest pool
> format while preserving its data. The existing containers should be
> accessible after the upgrade with the same level of functionality as
> before the upgrade. New features available at the pool, container or
> object level won’t be automatically enabled on existing pools unless
> upgrade is performed.

Reviewer: minor nitpicks
```suggestion
The pool upgrade operation upgrades a pool's disk format to the latest pool
format while preserving its data. The existing containers should be
accessible after the upgrade with the same level of functionality as
before the upgrade. New features available at the pool, container or
object level won’t be automatically enabled on existing pools unless
upgrade is performed.
```

> NB: The pool upgrade operation will evict any existing connections
> to this pool, pool will be unavailable to connect during pool upgrading.

Reviewer: ```suggestion
NB: The pool upgrade operation will evict any existing connections
to this pool - the pool will be unavailable to connect during pool upgrading.
```

To see all UpgradeNeeded pools:

```bash
$ dmg pool list
```

Below is an example of pool list

```bash
   Pool  Size   State Used Imbalance Disabled UpgradeNeeded?
   ----  ----   ----- ---- --------- -------- --------------
   pool1 1.0 GB Ready 0%   0%        0/4      1->2
   pool2 1.0 GB Ready 0%   0%        0/4      1->2
```

> For above example, pool1 and pool2 need be upgraded from pool

Reviewer: ```suggestion
For the above example, pool1 and pool2 need be upgraded from pool
```

> For the above example, pool1 and pool2 need be upgraded from pool

Reviewer 2: "need be upgraded from pool" --> "need to be upgraded from pool"

version 1 to pool version 2.

> NB: jump upgrading is not supported, eg upgrade pools created from
> DAOS v2.0 to DAOS v2.2 is ok, but not for DAOS v2.0 to v2.4 directly.

Reviewer: ```suggestion
NB: jump upgrading is not supported; e.g. upgrading pools created from
DAOS v2.0 to DAOS v2.2 is ok, but not for DAOS v2.0 to v2.4 directly.
```

   Pool performance domain (perf_domain)                                            root
   Tier placement policy (policy)                                                   type=io_size
   Pool redundancy factor (rd_fac)                                                  0
   Reclaim strategy (reclaim)                                                       lazy
   Performance domain affinity level of RP (rp_pda)                                 3
   Checksum scrubbing mode (scrub)                                                  off
   Checksum scrubbing frequency (scrub-freq)                                        604800
   Checksum scrubbing threshold (scrub-thresh)                                      0
   Self-healing policy (self_heal)                                                  exclude
   Rebuild space ratio (space_rb)                                                   0%
   Pool service replica list (svc_list)                                             [0]
   Pool service redundancy factor (svc_rf)                                          2
   Upgrade Status (upgrade_status)                                                  in progress
```

> Duration of upgrade depends on possible format change across different DAOS releases.
> It might trigger rebuild to re-place objects data which might be time-consuming.
> So `dmg pool upgrade` will return within seconds, but it might take hours to see Pool Status
> become 'in progress' to 'completed'. pool will keep offline and deny any pool connection until
> pool upgrade status become 'completed'.

Reviewer: ```suggestion
Duration of upgrade depends on possible format change across different DAOS releases.
It might trigger rebuild to re-place objects data which might be time-consuming.
So `dmg pool upgrade` will return within seconds, but it might take hours to see Pool Status
change from 'in progress' to 'completed'. The pool will stay offline and deny any pool connection until
pool upgrade status become 'completed'.
```

> pool upgrade status become 'completed'.

Reviewer 2: "become" --> "becomes"

> pool upgrade status becomes 'completed'.

Reviewer: @wangshilong If rebuild is triggered during upgrade - e.g. for EC data - is rebuild expected to finish before `upgrade_status=completed`? Or is it possible that the upgrade completes and rebuild is still ongoing?

Author: As document said, upgrade will be only marked as "completed" when rebuild finished.

Reviewer: I wanted to confirm because in my local testing I saw a case where upgrade was "completed" but rebuild was still ongoing. I'll try to reproduce that and create a ticket if needed. Thanks!

Author: Could we land this PR? DAOS-13941 has been fixed.
